### PR TITLE
Removed custom websocket implementation

### PR DESCRIPTION
**Pull request by davidfowl**
- Use the default websocket middleware
- Rewrote TestWebSocketConnectionFeature to use Channels instead of pipes
- The server basically copies what the client did (which isn't optimized at the moment)

### CODE REVIEW

File: src/Microsoft.AspNetCore.Sockets.Http/Transports/WebSocketsTransport.cs [hunk 1]
```diff
@@ -126,69 +117,93 @@ public WebSocketsTransport(WebSocketOptions options, IChannelConnection<Message>
...
+                    var truncBuffer = new ArraySegment<byte>(buffer.Array, 0, receiveResult.Count);
+                    incomingMessage.Add(truncBuffer);
+                    totalBytes += receiveResult.Count;
+                } while (!receiveResult.EndOfMessage);
 
-            // Write the message to the channel
-            return _application.Output.WriteAsync(message);
-        }
+                // Making sure the message type is either text or binary
+                Debug.Assert((receiveResult.MessageType == WebSocketMessageType.Binary || receiveResult.MessageType == WebSocketMessageType.Text), "Unexpected message type");
```
Comment (anurse) [inline src/Microsoft.AspNetCore.Sockets.Http/Transports/WebSocketsTransport.cs]: This isn't going to work for multi-part messages. Only the first message will have a `Binary` or `Text` message type, the remaining messages are of type `Continuation`. You need to capture the type of the first frame of the message for use here.

File: src/Microsoft.AspNetCore.Sockets.Http/Transports/WebSocketsTransport.cs [hunk 2]
```diff
@@ -126,69 +117,93 @@ public WebSocketsTransport(WebSocketOptions options, IChannelConnection<Message>
...
+                const int bufferSize = 4096;
+                var totalBytes = 0;
+                WebSocketReceiveResult receiveResult;
+                do
+                {
+                    var buffer = new ArraySegment<byte>(new byte[bufferSize]);
 
-            LogFrame("Receiving", frame);
+                    // Exceptions are handled above where the send and receive tasks are being run.
+                    receiveResult = await socket.ReceiveAsync(buffer, CancellationToken.None);
```
Comment (anurse) [inline src/Microsoft.AspNetCore.Sockets.Http/Transports/WebSocketsTransport.cs]: We should enforce that after the first frame of a message, all remaining frames **must** be of type `Continuation` or `Close`. Receiving `Text (Fin=false)` followed by `Text (Fin=true)` is a protocol error (the second message must be `Continuation`) and we should close the socket when that occurs, with status code `1002` (Protocol Error)

File: src/Microsoft.AspNetCore.Sockets.Http/Transports/WebSocketsTransport.cs [hunk 3]
```diff
@@ -126,69 +117,93 @@ public WebSocketsTransport(WebSocketOptions options, IChannelConnection<Message>
...
+            while (true)
             {
-                return Task.CompletedTask;
-            }
+                const int bufferSize = 4096;
+                var totalBytes = 0;
+                WebSocketReceiveResult receiveResult;
+                do
+                {
+                    var buffer = new ArraySegment<byte>(new byte[bufferSize]);
```
Comment (anurse) [inline src/Microsoft.AspNetCore.Sockets.Http/Transports/WebSocketsTransport.cs]: I assume we'll want to pool this memory eventually.

File: src/Microsoft.AspNetCore.Sockets.Http/Transports/WebSocketsTransport.cs [hunk 4]
```diff
@@ -126,69 +117,93 @@ public WebSocketsTransport(WebSocketOptions options, IChannelConnection<Message>
...
-                                opcode: _lastFrameIncomplete ? WebSocketOpcode.Continuation : opcode,
-                                payload: ReadableBuffer.Create(message.Payload));
+                            var messageType = message.Type == MessageType.Binary ?
+                                WebSocketMessageType.Binary :
+                                WebSocketMessageType.Text;
 
-                            _lastFrameIncomplete = !message.EndOfMessage;
+                            if (_logger.IsEnabled(LogLevel.Debug))
+                            {
+                                _logger.LogDebug(
```
Comment (anurse) [inline src/Microsoft.AspNetCore.Sockets.Http/Transports/WebSocketsTransport.cs]: Use the logging syntax that maintains structured log messages: `_logger.LogDebug("Sending frame: Fin={fin}, Payload={length} bytes", message.EndOfMessage, message.Payload.Length);`

I wish we could some how tap in to `FormattedString` to make this work while still preserving structured logging :)

File: src/Microsoft.AspNetCore.Sockets.Http/Transports/WebSocketsTransport.cs [hunk 5]
```diff
@@ -126,69 +117,93 @@ public WebSocketsTransport(WebSocketOptions options, IChannelConnection<Message>
...
+                    message = new Message(messageBuffer, messageType, receiveResult.EndOfMessage);
+                }
+                else
+                {
+                    var buffer = new byte[incomingMessage[0].Count];
+                    Buffer.BlockCopy(incomingMessage[0].Array, incomingMessage[0].Offset, buffer, 0, incomingMessage[0].Count);
+                    message = new Message(buffer, messageType, receiveResult.EndOfMessage);
+                }
+
+                _logger.LogInformation("Passing message to application. Payload size: {0}", message.Payload.Length);
```
Comment (anurse) [inline src/Microsoft.AspNetCore.Sockets.Http/Transports/WebSocketsTransport.cs]: Give `0` a name: `{length}`

File: src/Microsoft.AspNetCore.Sockets.Http/Transports/WebSocketsTransport.cs [hunk 6]
```diff
@@ -126,69 +117,93 @@ public WebSocketsTransport(WebSocketOptions options, IChannelConnection<Message>
...
-            var effectiveOpcode = frame.Opcode;
-            if (frame.Opcode == WebSocketOpcode.Continuation)
-            {
-                effectiveOpcode = _lastOpcode;
-            }
-            else
-            {
-                _lastOpcode = frame.Opcode;
-            }
+                    _logger.LogDebug("Message received. Type: {messageType}, size: {size}, EndOfMessage: {endOfMessage}",
```
Comment (anurse) [inline src/Microsoft.AspNetCore.Sockets.Http/Transports/WebSocketsTransport.cs]: See, you do it right here :).

File: src/Microsoft.AspNetCore.Sockets.Http/Transports/WebSocketsTransport.cs [hunk 7]
```diff
@@ -88,23 +79,23 @@ public WebSocketsTransport(WebSocketOptions options, IChannelConnection<Message>
 
                 // Shutting down because we received a close frame from the client.
                 // Complete the input writer so that the application knows there won't be any more input.
-                _logger.LogDebug("Client closed connection with status code '{0}' ({1}). Signaling end-of-input to application", receiving.Result.Status, receiving.Result.Description);
+                _logger.LogDebug("Client closed connection with status code '{0}' ({1}). Signaling end-of-input to application", receiving.Result.CloseStatus, receiving.Result.CloseStatusDescription);
```
Comment (anurse) [inline src/Microsoft.AspNetCore.Sockets.Http/Transports/WebSocketsTransport.cs]: I know this was already wrong when you started, but update this one to use named format specifiers too.

**Comment by BrennanConroy** [comment]
Remove `pip install` from travis.yml!

File: src/Microsoft.AspNetCore.Sockets.Http/Transports/WebSocketsTransport.cs (hunk 1 above)
Comment (moozzyk) [inline src/Microsoft.AspNetCore.Sockets.Http/Transports/WebSocketsTransport.cs]: `WebSocketMessageType` enum [does not have](https://msdn.microsoft.com/en-us/library/system.net.websockets.websocketmessagetype(vs.110).aspx) the `Continuation` member...

File: src/Microsoft.AspNetCore.Sockets.Http/Transports/WebSocketsTransport.cs (hunk 2 above)
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Sockets.Http/Transports/WebSocketsTransport.cs]: Already works

**Pull request merged**